### DISCUSSION TITLE
SQLAlchemy 2.0 models declared with DeclarativeBase migration not created (on sqlite) by flask-migrate

### DISCUSSION THREAD

**Discussion by alessandromannini** [miguelgrinberg/Flask-Migrate#509] [Q&A] [answered]
Hi,

it seems that the env.py generated by flask-migrate isn't able to generate migrations if models are described using the SQLAlchemy 2.0-style (with DeclarativeBase) class.

I defined some model (see attached txt) using the new declarative syntax of SQLAlchemy and tried to generate the first migration throught Flask-Migrate

 but I receive the message "No changes detected".

I've tried many solutions without success.
As suggested in the comment I also tried to alter env.py in this way:
`# add your model's MetaData object here`
`# for 'autogenerate' support`
`# from myapp import mymodel`
`# target_metadata = mymodel.Base.metadata`
`config.set_main_option('sqlalchemy.url', current_app.config.get('SQLALCHEMY_DATABASE_URI'))`
`target_db = current_app.extensions['migrate'].db`
**`from application import models`**
**`target_metadata = models.Base.metadata`**

Notice that If I put an "old style" (with Column) declaration in my models a **create** migration is generated.
I tried to printf target_metadata.tables.values() and it is correctly filled but i also noticed that the env.py script make use of the function get_metadata() that **ignore** the target_metadata set above....
Finally i've tested that if i set:
**`MY_target_metadata`**` = models.Base.metadata`

and, below, i change the context.configure that way:

...[CUT]...
`   with connectable.connect() as connection`:
`        context.configure(`
`            connection=connection`,
`            target_metadata=`**`MY_target_metadata`**`,`
`            process_revision_directives=process_revision_directives,`
`            **current_app.extensions['migrate'].configure_args`
`        )`

the first migration is generated.

See discussion with the Alembic team [here](https://github.com/sqlalchemy/alembic/discussions/1189).

Thanks
Alessandro

[models.txt](https://github.com/miguelgrinberg/Flask-Migrate/files/10847328/models.txt)

**Comment by miguelgrinberg** [comment]
I'm going to copy the answer I put in the Alembic discussion here as well:

Just so that we are all on the same page, please keep in mind that Flask-Migrate requires the Flask-SQLAlchemy or Alchemical extensions for Flask, and it provides its own customized template that integrate with these extensions. To work directly with SQLAlchemy as the OP is intending, the env.py file would need to be changed and made to look more like the official Alembic template, but I would suggest using Alembic directly in that case.

**Reply by alessandromannini** [reply]
Hi Miguel,

I'm using Flask-Migrate + Flask-SQLAlchemy... This is from my requirements:
`Flask==2.2.3`
`Flask-Migrate==4.0.4`
`Flask-SQLAlchemy==3.0.3`

and I try to generate the first migration with:
`flask db init` (that works and generate the enpty db and the customized env.py)
`flask db migrate -m "Tables creation"` (that respond No changes detected)

As I said I tried to modify the env.py as suggested in the comment but it doesn't work.
It is a new development framework for me so maybe I'm wrong but.... don't understand where.
Thanks

Alessandro

**Reply by miguelgrinberg** [reply]
What I meant is that you are using your own declarative base, not Flask-SQLAlchemy's. You aren't really using Flask-SQLAlchemy as far as I can see. There is no reference to it in your models file.

**Comment by alessandromannini** [comment, marked as answer]
Ok! I understand now.
Thx
A

For flask-newbie like me.

Models must be subclassed from the Model class of **Flask-SQLAlchemy** library not from the (generic) SQLAlchemy DeclarativeBase class.
In my case db is the Flask-SQLAlchemy object and I have corrected in:

`class Layout(db.Model):......`
`class Routes(db.Model):......`
`class Nodes(db.Model):......`

If you need to customize, you have to subclass Model and instanciate SQLAlchemy with the model_class parameter like this:

`class MyBase(Model):......`
`db = SQLAlchemy(model_class=MyBase):.....`
'class Layout(db.Model)'

as documented in the [advanced customization](https://flask-sqlalchemy.palletsprojects.com/en/3.0.x/customizing/).
